reference/test/matrix/sparsity_csr_kernels.cpp

```diff
@@ -0,0 +1,407 @@
+    std::unique_ptr<Mtx> mtx3_sorted;
+    std::unique_ptr<Mtx> mtx3_unsorted;
+};
+
+
+TEST_F(SparsityCsr, AppliesToDenseVector)
+{
+    auto x = gko::initialize<Vec>({2.0, 1.0, 4.0}, exec);
+    auto y = Vec::create(exec, gko::dim<2>{2, 1});
+
+    mtx->apply(x.get(), y.get());
+
+    EXPECT_EQ(y->at(0), 7.0);
+    EXPECT_EQ(y->at(1), 1.0);
+}
+
+
+TEST_F(SparsityCsr, AppliesToDenseMatrix)
+{
+    auto x = gko::initialize<Vec>({{2.0, 3.0}, {1.0, -1.5}, {4.0, 2.5}}, exec);
+    auto y = Vec::create(exec, gko::dim<2>{2});
+
+    mtx->apply(x.get(), y.get());
+
+    EXPECT_EQ(y->at(0, 0), 7.0);
+    EXPECT_EQ(y->at(1, 0), 1.0);
+    EXPECT_EQ(y->at(0, 1), 4.0);
+    EXPECT_EQ(y->at(1, 1), -1.5);
+}
+
+
+TEST_F(SparsityCsr, AppliesLinearCombinationToDenseVector)
+{
+    auto alpha = gko::initialize<Vec>({-1.0}, exec);
+    auto beta = gko::initialize<Vec>({2.0}, exec);
+    auto x = gko::initialize<Vec>({2.0, 1.0, 4.0}, exec);
+    auto y = gko::initialize<Vec>({1.0, 2.0}, exec);
+
+    mtx->apply(alpha.get(), x.get(), beta.get(), y.get());
+
+    EXPECT_EQ(y->at(0), -5.0);
+    EXPECT_EQ(y->at(1), 3.0);
+}
+
+
+TEST_F(SparsityCsr, AppliesLinearCombinationToDenseMatrix)
+{
+    auto alpha = gko::initialize<Vec>({-1.0}, exec);
+    auto beta = gko::initialize<Vec>({2.0}, exec);
+    auto x = gko::initialize<Vec>({{2.0, 3.0}, {1.0, -1.5}, {4.0, 2.5}}, exec);
+    auto y = gko::initialize<Vec>({{1.0, 0.5}, {2.0, -1.5}}, exec);
+
+    mtx->apply(alpha.get(), x.get(), beta.get(), y.get());
+
+    EXPECT_EQ(y->at(0, 0), -5.0);
+    EXPECT_EQ(y->at(1, 0), 3.0);
+    EXPECT_EQ(y->at(0, 1), -3.0);
+    EXPECT_EQ(y->at(1, 1), -1.5);
+}
+
+
+TEST_F(SparsityCsr, ApplyFailsOnWrongInnerDimension)
+{
+    auto x = Vec::create(exec, gko::dim<2>{2});
+    auto y = Vec::create(exec, gko::dim<2>{2});
+
+    ASSERT_THROW(mtx->apply(x.get(), y.get()), gko::DimensionMismatch);
+}
+
+
+TEST_F(SparsityCsr, ApplyFailsOnWrongNumberOfRows)
+{
+    auto x = Vec::create(exec, gko::dim<2>{3, 2});
+    auto y = Vec::create(exec, gko::dim<2>{3, 2});
+
+    ASSERT_THROW(mtx->apply(x.get(), y.get()), gko::DimensionMismatch);
+}
+
+
+TEST_F(SparsityCsr, ApplyFailsOnWrongNumberOfCols)
+{
+    auto x = Vec::create(exec, gko::dim<2>{3});
+    auto y = Vec::create(exec, gko::dim<2>{2});
+
+    ASSERT_THROW(mtx->apply(x.get(), y.get()), gko::DimensionMismatch);
+}
+
+
+TEST_F(SparsityCsr, SquareMtxIsTransposable)
+{
+    // clang-format off
+   auto mtx2 = gko::initialize<gko::matrix::SparsityCsr<>>(
+               {{1.0, 1.0, 1.0},
+                {0.0, 1.0, 0.0},
+                {0.0, 1.0, 1.0}}, exec);
+    // clang-format on
+
+    auto trans = mtx2->transpose();
+    auto trans_as_sparsity =
+        static_cast<gko::matrix::SparsityCsr<> *>(trans.get());
+
+    // clang-format off
+   GKO_ASSERT_MTX_NEAR(trans_as_sparsity,
+                   l({{1.0, 0.0, 0.0},
+                      {1.0, 1.0, 1.0},
+                      {1.0, 0.0, 1.0}}), 0.0);
+    // clang-format on
+}
```

Comment:
The generic `LinOp` derived stuff even with some constructors is in the `core/test/matrix/` , whereas anything that needs the constructor that uses the `copy_and_convert_to` to directly initialize the matrix from an input dense/CSR or any other matrix that implements a `convert_to` should go in the `reference/test/matrix/` because `convert_to` is a kernel and cannot be called with the core module and as the constructor is not technically a kernel even though internally it does call the `copy_and_convert_to` it is in the `reference/test/matrix/sparsity_csr.cpp` file.

Reply:
I thought we use `_kernel` when we directly call kernels, but that was never really a rule (and we also have `_kernel` test files which call class functions).
All in all, I think this is fine the way it is, I was just confused why you had a pretty much empty test file, and almost all tests inside the `_kernel` file.